### PR TITLE
Add following configs for network tuning.

### PR DESCRIPTION
**Pull request by zzmao**
routerConfig.routerMaxPutChunkNumber
networkConfig.selectorMaxRunningTimeInMs
networkConfig.selectorMaxKeyToProcess

### CODE REVIEW

File: ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java [hunk 1]
```diff
@@ -111,6 +111,13 @@
   @Default("4*1024*1024")
   public final int routerMaxPutChunkSizeBytes;
 
+  /**
+   * The maximum number of chunks can be used at the same time for a PutOperation
+   */
+  @Config("router.max.put.chunk.number")
+  @Default("4")
+  public final int routerMaxPutChunkNumber;
```
Comment (cgtz) [inline ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java]: I ended up adding this config in another PR, so you can remove this

File: ambry-network/src/test/java/com.github.ambry.network/SelectorTest.java [hunk 2]
```diff
@@ -55,8 +58,12 @@ public SelectorTest(int poolSize) {
...
     this.server = new EchoServer(18283);
     this.server.start();
-    this.selector = new Selector(new NetworkMetrics(new MetricRegistry()), SystemTime.getInstance(), null,
-        selectorExecutorPoolSize);
+    Properties props = new Properties();
+    props.setProperty("selector.executor.pool.size", Integer.toString(selectorExecutorPoolSize));
+    VerifiableProperties vprops = new VerifiableProperties(props);
+    NetworkConfig networkConfig = new NetworkConfig(vprops);
+    this.selector =
+        new Selector(new NetworkMetrics(new MetricRegistry()), SystemTime.getInstance(), null, networkConfig);
```
Comment (cgtz) [inline ambry-network/src/test/java/com.github.ambry.network/SelectorTest.java]: is there a way to test the new configs?

File: ambry-network/src/main/java/com.github.ambry.network/Selector.java [hunk 3]
```diff
@@ -340,6 +346,10 @@ private void pollOnMainThread(long timeoutMs, List<NetworkSend> sends) throws IO
       metrics.selectorReadyKeyCount.inc(keys.size());
       Iterator<SelectionKey> iter = keys.iterator();
       while (iter.hasNext()) {
+        if (networkConfig.selectorMaxRunningTimeInMs != -1
```
Comment (cgtz) [inline ambry-network/src/main/java/com.github.ambry.network/Selector.java]: It seems like one of these configs applies only for main thread polling and one applies for executor pool polling

**Comment by zzmao** [comment]
PR updated. Feel free to review. Working on a test per @cgtz 's comment, will push it soon.

File: ambry-network/src/main/java/com.github.ambry.network/SocketServer.java [hunk 4]
```diff
@@ -66,6 +67,7 @@
   private SSLFactory sslFactory;
 
   public SocketServer(NetworkConfig config, SSLConfig sslConfig, MetricRegistry registry, ArrayList<Port> portList) {
+    this.networkConfig = config;
     this.host = config.hostName;
     this.port = config.port;
     this.numProcessorThreads = config.numIoThreads;
```
Comment (jsjtzyy) [inline ambry-network/src/main/java/com.github.ambry.network/SocketServer.java]: `selectorExecutorPoolSize` can be removed from this class.

Comment (zzmao) [reply ambry-network/src/main/java/com.github.ambry.network/SocketServer.java]: addressed.

File: ambry-network/src/main/java/com.github.ambry.network/Selector.java [hunk 5]
```diff
@@ -495,12 +523,17 @@ private void pollWithExecutorPool(long timeoutMs, List<NetworkSend> sends) throw
...
-      for (SelectionKey keyWithError : readWriteKeySet) {
+      // Keys hit exception should be closed.
+      for (SelectionKey keyWithError : processingKeySet) {
         close(keyWithError);
       }
       checkUnreadyConnectionsStatus();
-      metrics.selectorIOCount.inc();
-      metrics.selectorIOTime.update(time.milliseconds() - endSelect);
+      long selectorIOTime = time.milliseconds() - endSelect;
+      if (selectorIOTime >= 50) {
```
[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-network/src/main/java/com.github.ambry.network/Selector.java]: Yes, it's based on observation to filter out small selectorIOTime that we don't care about. 
`pollOnMainThread` also do the same thing for `selectorIOTime`. Prefer not to change it now.

[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-network/src/main/java/com.github.ambry.network/Selector.java]: `checkUnreadyConnectionsStatus` is just some local computation. It should be a fast call.

**Comment by zzmao** [comment]
@cgtz  Good to merge this?

**Pull request merged**